### ISSUE TITLE
refactor the OPT framework using C++11 and openMP

### ISSUE THREAD

**Issue by yorkerlin** [shogun-toolbox/shogun#3316]
Project https://github.com/yorkerlin/TMAC is our starting point.
This project is written in C++11 using template.
What is more, coordinate descend methods are missing in the existing shogun's OPT framework.

**Comment by karlnapf** [comment]
Great!

**Comment by yorkerlin** [comment]
any comment about using openMP 4.0 since openMP 4.0 supports GPU?

**Comment by yorkerlin** [comment]
or c++ build-in thread lib since the lib is provided in c++11

**Comment by karlnapf** [comment]
we want to use linalg for gpu stuff.
c++11 should happen, but not before end of GSoC 2016 please.

**Comment by yorkerlin** [comment]
should we use openMP or C++11 for `task based concurrency` ?

**Comment by yorkerlin** [comment]
`pthread` and `openMP` are used in existing code base.
We may use the built-in thread lib in C++11 to replace `pthread` if we want to support for windows in the long run.

**Comment by karlnapf** [comment]
I guess that since we are now switching to c++11, we can make use of the in built parallel of the language.
omp doesnt work for windows?

**Comment by yorkerlin** [comment]
`openMP` should be fine for windows.
I am not sure whether `pthread` is  supported in windows without `cygwin`.

**Comment by karlnapf** [comment]
Not sure either. omp is the easiest thing to use. But we cannot stop threads with it.
C++11 might be better there, maybe send a patch using it?

**Comment by yorkerlin** [comment]
`openMP` should be fine if thread does not throw exception. 
My plan is to replace `pthread` in GP with the thread lib in `C++11`

**Comment by karlnapf** [comment]
yeah do it! BTW I also talked to Michaelis Titsias and he said he would have a look at our code and check where it can be improved, will write an email

**Comment by vigsterkr** [comment]
but openmp actually uses `pthreads` so.....? :)

**Comment by yorkerlin** [comment]
@karlnapf 
some empirical results about  large scale kernel learning using coordinate descent
http://arxiv.org/abs/1602.05310

**Comment by yorkerlin** [comment]
@vigsterkr 
Well, if a openMP compiler choose to use pthread, it should be fine. :)
My point is we should use openMP or/and the thread lib in c++11 at least in new projects.

**Comment by yorkerlin** [comment]
reference:
https://github.com/mlpack/mlpack/tree/master/src/mlpack/core/optimizers
